Sources/SwiftCuckoo/Session.swift

```diff
@@ -0,0 +1,90 @@
+///
+/// The `Session` structure is designed to encapsulate the details of a task that
+/// includes its start time, optional end time, and the duration calculated
+/// from these timestamps. This structure aids in managing and reporting
+/// the time spent on various tasks, providing valuable insights for users
+/// looking to optimize their productivity.
+public struct Session: Equatable, Hashable {
+    /// A unique identifier for the session instance.
+    public var id: Identifier
+    
+    /// The start time of the session. This indicates when the session began.
+    public var startTime: Date
+    
+    /// The optional end time of the session. This indicates when the session
+    /// was completed or could still be ongoing.
+    public var endTime: Date?
+
+    /// Initializes a new session instance with a unique identifier, a start
+    /// time, and an optional end time.
+    ///
+    /// - Parameters:
+    ///   - id: A unique identifier for the session. This helps in managing
+    ///          and tracking individual session instances.
+    ///   - startTime: The starting time of the session, indicating when it begins.
+    ///   - endTime: The end time of the session, representing when the session is
+    ///              completed. This parameter is optional and defaults to nil for ongoing sessions.
+    ///
+    /// - Note: Ensure that the `endTime` is provided if the session is completed.
+    package init(id: Identifier, startTime: Date, endTime: Date? = nil) {
+        self.id = id
+        self.startTime = startTime
+        self.endTime = endTime
+    }
+    
+    /// Calculates the duration of the session.
+    ///
+    /// - Throws: An error of type `Error` if:
+    ///   - The `endTime` is missing.
+    ///   - The `endTime` is before the `startTime`, indicating an invalid time range.
+    ///
+    /// - Returns: The duration of the session as a `TimeInterval` representing
+    ///   the time spent from the start time to the end time.
+    public func duration() throws -> TimeInterval {
+        guard let endTime else {
+            throw Error.missingEndTime
+        }
+        
+        guard endTime >= startTime else {
+            throw Error.endTimeBeforeStartTime
+        }
+        
+        return endTime.timeIntervalSince(startTime)
+    }
+}
+
+// MARK: - Identifier
+
+extension Session {
+    /// A unique identifier for the session, conforming to `RawRepresentable`,
+    /// `Equatable`, and `Hashable` protocols for easy management and comparison.
+    public struct Identifier: RawRepresentable, Equatable, Hashable {
+        /// The raw string value representing the identifier.
+        public let rawValue: String
+        
+        /// Initializes a new identifier with a given raw value.
+        ///
+        /// - Parameter rawValue: The string value that represents the identifier.
+        public init(rawValue: String) {
+            self.rawValue = rawValue
+        }
+    }
```

Comment:
ℹ️ This `Identifier` approach is similar to `Notification.Name`.

So instead of using `String` directly to identify a `Session` it uses this structure and the advantage of it is that the client can add auto-completion possibility.

## Example: 

### Without auto-completion:
```swift
let session = Session(id: Session.Identifier(rawValue: "example_id"), ...)
```

### with auto-completion
```swift
extension Session.Identifier {
    static let exampleId = Session.Identifier(rawValue: "example_id")
}

let session = Session(id: .exampleId, ...)
```

On the **SessionTests.swift** has an example:
<img width="538" alt="Screenshot 2024-09-03 at 06 00 11" src="https://github.com/user-attachments/assets/76e4a687-b035-4614-a1bd-2f40b47b0f2f">


I like this approach because it avoids typo errors.

Reply:
This is clever